README
NAME
    HTML::Template::Compiled - Template System Compiles HTML::Template files
    to Perl code

VERSION
    $VERSION = "1.001_001"

SYNOPSIS
      use HTML::Template::Compiled;
      # recommended options:
      # case_sensitive => 1
      # search_path_on_include => 1
      # use_query => 0
      # default_escape => 'HTML' # <-- HIGHLY RECOMMENDED

      # note that the following
      # use HTML::Template::Compiled speed => 1
      # is deprecated (can be problematic under persistent environments)

      # or for the biggest compatibility with HTML::Template
      # case_sensitive => 0
      # search_path_on_include => 0
      # use_query => 1
      # note that the following
      # use HTML::Template::Compiled compatible => 1;
      # is deprecated (can be problematic under persistent environments)

      # or use HTML::Template::Compiled::Classic

        my $htc = HTML::Template::Compiled->new(
            filename            => 'test.tmpl',
            case_sensitive      => 1,
            default_escape      => 'HTML',
        );
        $htc->param(
            BAND => $name,
            ALBUMS => [
                { TITLE => $t1, YEAR => $y1 },
                { TITLE => $t2, YEAR => $y2 },
            ],
        );
        print $htc->output;

      test.tmpl:
      Band: <TMPL_VAR BAND>
      <TMPL_LOOP ALBUMS>
      Title: <TMPL_VAR TITLE> (<TMPL_VAR YEAR>)
      </TMPL_LOOP>

      Or use different tag styles:
      Band: <%= BAND %>
      <%loop ALBUMS %>
      Title: <%= TITLE %> (<%= YEAR %>)
      <%/loop %>
      Band: [%= BAND %]
      [%loop ALBUMS %]
      Title: [%= TITLE %] ([%= YEAR %])
      [%/loop %]

DESCRIPTION
    HTML::Template::Compiled is a template system which can be used for
    HTML::Template templates with almost the same API. It offers more
    flexible template delimiters, additional tags and features, and by
    compiling the template into perl code it can run significantly faster in
    persistent environments such as FastCGI or mod_perl.

    The goal is to offer more features for flexibility but keep the basic
    syntax as easy as it is.

    Features at a glance:

    Dot notation for objects, hashes and arrays
    Use expressions without any disadvantages like those in
    HTML::Template::Expr
    Write escaping plugins and plugins for new tags
    Alternate delimiters, e.g. "[%if %]" and "<%if %>"
    Avoid "global_vars" option by using the "SET_VAR" tag to create aliases.
    Tags ELSIF, EACH, WHILE, COMMENT, WRAPPER, SWITCH/CASE, INCLUDE_VAR
    Chomp newlines/whitespace

    For a quick reference, see HTML::Template::Compiled::Reference.

    As the basic features work like in HTML::Template, please get familiar
    with its documentation before.

    HTML::Template::Compiled (HTC) does not implement all features of
    HTML::Template (see "COMPATIBILITY"), and it has got some additional
    features which are explained below: "ADDITIONAL FEATURES"

    See "BENCHMARKS" for some examples on the performance. Since it depends
    highly on the options used and on the template size there can be no
    general statement on its performance.

    You might want to use HTML::Template::Compiled::Lazy for CGI
    environments as it doesn't parse the template before calling output. But
    note that HTC::Lazy isn't much tested, and I don't use it myself, so
    there's a lack of experience. If you use it and have problems, please
    report.

    HTC will use a lot of memory because it keeps all template objects in
    memory. If you are on mod_perl, and have a lot of templates, you should
    preload them at server startup to be sure that it is in shared memory.
    At the moment HTC is not fully tested for keeping all data in shared
    memory (e.g. when a copy-on-write occurs), but it seems like it's
    behaving well. For preloading you can use
    HTML::Template::Compiled->preload($cache_dir).

    Generating code, writing it on disk and later eval() it can open
    security holes, for example if you have more users on the same machine
    that can access the same files (usually an http server running as 'www'
    or 'nobody'). See "SECURITY" for details what you can do to safe
    yourself.

    NOTE: If you don't need any of the additional features listed below and
    if you don't need the speed (in many cases it's probably not worth
    trading speed for memory), then you might be better off with just using
    HTML::Template.

    NOTE2: If you have any questions, bug reports, send them to me and not
    to Sam Tregar. This module is developed by me at the moment,
    independently from HTML::Template, although I try to get most of the
    tests from it passing for HTC. See "RESOURCES" for current information.

  FEATURES FROM HTML::TEMPLATE
    TMPL_VAR
    TMPL_LOOP
    TMPL_(IF|UNLESS|ELSE)
    TMPL_INCLUDE
    HTML_TEMPLATE_ROOT
    ESCAPE=(HTML|URL|JS|0)
    DEFAULT=...
    "__first__", "__last__", "__inner__", "__outer__", "__odd__",
    "__counter__", "__even__"
    <!-- TMPL_VAR NAME=PARAM1 -->
    case insensitive var names
        use option case_sensitive => 0 to use this feature (slow down)

    filters
    vars that are subrefs - not implemented, only in
    HTML::Template::Compiled::Classic
    scalarref, arrayref, filehandle
    "global_vars"
    "query"
        Has a bug (doesn't return parameters in included files of included
        files). I'm working on that.

  ADDITIONAL FEATURES
    What can HTC do for you additionally to HTML::Template?

    tag TMPL_ELSIF
        No need to have cascading "if-else-if-else"s

    tag TMPL_EACH
        Iterate over a hash. See "TMPL_EACH"

    tag TMPL_WITH
        see "TMPL_WITH"

    tag TMPL_WHILE
        see "TMPL_WHILE"

    tag TMPL_SET_VAR
        see "SET_VAR"

    tag TMPL_USE_VARS
        see "USE_VARS"

    tags TMPL_COMMENT, TMPL_NOPARSE, TMPL_VERBATIM
        see "TMPL_COMMENT", "TMPL_NOPARSE", "TMPL_VERBATIM"

    tag TMPL_WRAPPER
        see "WRAPPER"

    "__index__"
        Additional loop variable ("__counter__ -1")

    "__break__"
        Additional loop variable (see "TMPL_LOOP")

    "__filename__", "__filenameshort__" (since 0.91_001)
        Insert the template filename for debugging:

            <%= __filename__ %>
            <%= __filenameshort__ %>

        will turn out as: templates/path/file.html path/file.html

        See also option debug_file in "OPTIONS" for adding the filename
        globally.

    tags TMPL_SWITCH, TMPL_CASE
        see "TMPL_SWITCH"

    "TMPL_PERL"
        Include perl code in your template. See "RUNNING PERL WITH
        TMPL_PERL"

    CHOMP
        New in version 0.96_001, please report any bugs and send me
        suggestions.

        You can set global chomp options in the constructor. These work like
        in Template-Toolkit:

            my $htc = HTML::Template::Compiled->new(
                pre_chomp  => 0, # 0, 1, 2, 3, default 0
                post_chomp => 1, # 0, 1, 2, 3, default 0
            );

        Meaning of the values: 0: Don't chomp 1: remove only spaces in the
        line before or after the tag 2: remove all whitespaces before or
        after the tag, and replace with one space 3: remove all whitespaces
        before or after the tag

        In the template you can change that feature by using PRE_CHOMP and
        POST_CHOMP attributes:

            <%= foo PRE_CHOMP=3 POST_CHOMP=1 %>

        The experimental tags +..._chomp have been removed.

    Generating perl code
        See "IMPLEMENTATION"

    better variable access
        dot-notation for accessing hash values. See "EXTENDED VARIABLE
        ACCESS"

    rendering objects
        dot-notation for accessing object methods. See "RENDERING OBJECTS"

    output to filehandle
        See "OPTIONS"

    Dynamic includes
        "INCLUDE_VAR", "INCLUDE_STRING". See "INCLUDE"

    tag TMPL_IF_DEFINED
        Check for definedness instead of truth: <TMPL_IF_DEFINED NAME="var">

    ALIAS
        Set an alias for a loop variable. You can use the alias then with
        $alias. The syntax without the "$" is also possible but not
        recommended any more.

        For example, these two loops are functionally equivalent:

         <tmpl_loop foo>
           <tmpl_var _>
         </tmpl_loop foo>
         <tmpl_loop foo alias=current>
           <tmpl_var $current>
         </tmpl_loop foo>

        This works with "TMPL_LOOP" and "TMPL_WHILE" at the moment.

        You can also set aliases with the "SET_VAR" tag. See "SET_VAR"

        To use template parameters with a "$" at the beginning (which is not
        officially supported, but some are obviously using it), you can set:

            local $HTML::Template::Compiled::Compiler::DISABLE_NEW_ALIAS = 1;

        This is only a temporary workaround and will be removed some day!

        Note that you are also able to access variables with dollar signs
        like this:

            <tmpl_var _.$foo >

        since underscore means current position in the parameter stash, and
        aliases are only recognized at the beginning of a template var. But
        note that dollar signs are still not officially supported.

    Chained escaping
        See "ESCAPING"

    tagstyles
        For those who like it (i like it because it is shorter than TMPL_),
        you can use <% %> tags and the <%= tag instead of <%VAR (which will
        work, too):

         <%IF blah%>  <%= VARIABLE%>  <%/IF%>

        Define your own tagstyles and/or deactivate predefined ones. See
        "OPTIONS" tagstyle.

    pre_chomp, post_chomp
        See "CHOMP"

  MISSING AND DIFFERENT FEATURES
    There are some features of H::T that are missing or behaving different.
    I'll try to list them here.

   MISSING FEATURES
    die_on_bad_params
        I don't think I'll implement that.

    force_untaint
        Not planned at the moment

    vanguard_compatibility_mode
        Not planned.

    shared_cache, double_cache
        Not planned at the moment

    blind_cache
        Not sure if I should implement. In HTC you have the possibility to
        set the expire time of the templates (after that time in memory the
        template file is rechecked if it has changed), so setting a very
        high value for expire_time would have the same effect as
        blind_cache. See "CACHING" "expire_time"

    double_file_cache
        If I understand correctly, in HT, this enables memory and file cache
        at the same time. In HTC, this is not needed. If you use file_cache
        and cache, both are used.


Changes
Revision history for Perl module HTML::Template::Compiled.

1.002_001 Fri Apr  8 21:23:29 CEST 2016
- Bugfix: prevent unexpcted Data::Dumper config to mess with the output
  (github issue #10)
- Fixed typos (github PR #9)
- Storable >= 2.05 (CODE refs)

1.002 Sun Jul 26 01:43:26 CEST 2015
- Regular release, see developer versions

1.001_002 Tue Jul 21 20:48:37 CEST 2015
- Bugfix: Issues with global_vars #6, #7 (rpv-tomsk++)
- Changed to Dist::Zilla

1.001_001 Tue May 21 00:50:22 CEST 2013
- New option: optimize
- Minor optimization

1.001 Sat May  4 12:49:44 CEST 2013
- Regular release, see developer versions for changes

1.000_006 Fri Jan 25 22:57:24 CET 2013
- Minor optimizations for file cache
- Remove ::Plugin::DHTML (own distribution)

1.000_004 Thu Jan 10 00:29:59 CET 2013
- New features: added options warnings and line_info

1.000_003 Mon Dec 31 19:05:54 CET 2012
- fix test

1.000_002 Sat Dec  1 21:06:57 CET 2012
- make parallel testing possible
- remove ::Plugin::NumberFormat (has its own distribution now)

1.000_001 Tue Nov 27 19:55:00 CET 2012
- Minor optimizations
- New Feature: attribute sortby for TMPL_EACH

1.000 Tue Nov 13 20:45:44 CET 2012
- regular release

0.99_001 Sat Aug  4 17:32:03 CEST 2012
- Bugfix: reloading changed includes didn't always work

0.99 Mon Jul  2 21:48:39 CEST 2012
- fixed tests (deprecation in perl 5.17 and Data::TreeDumper prereq)
- also see developer version changes

0.98_003 Sun Jun 10 19:25:15 CEST 2012
- removed test cache directory from distribution
- added some modules to recommends

0.98_002 Wed Jun  6 20:43:05 CEST 2012
- Workaround for using var names with dollars:
  local $HTML::Template::Compiled::Compiler::DISABLE_NEW_ALIAS = 1;

0.98_001 Mon Jun  4 23:53:45 CEST 2012
- require perl 5.8.1
- Hotfix for include_var and file caching

0.98 Mon Jun  4 18:18:40 CEST 2012
- IMPORTANT: Change: use aliases (alias=..., SET_VAR) via $aliasname.
  USE_VARS will be removed in the future. Old alias syntax still works
  for now.

0.97_006 Sun Jun  3 19:26:50 CEST 2012
- new plugin HTML::Template::Compiled::Plugin::NumberFormat

0.97_005 Sat Jun  2 12:47:06 CEST 2012
- New feature: option cache_debug
- New feature: TMPL_WRAPPER
- Another bugfix for search_path_on_include

0.97_004 Sun May 20 18:38:32 CEST 2012
- Change: Implemented searching the path and search_path_on_include like
  in HTML::Template (finally) and add an additional value. PLEASE TEST!
- Improved file cache performance
- Bugfix: old bug with invalid tags in noparse, verbatim, comment
- Remove old Storable workaround (0.94_002)

0.97_001 Thu May 17 14:54:30 CEST 2012
- fix HTML::Template::Compiled::Classic: allow every character in template
  var names (RT 70676)
- New option expire_time instead of global HTML::Template::Compiled->ExpireTime
- Added loop context vars __even__, __outer__
- Change: removed documentation of class methods ->CaseSensitive,
  ->SearchPathOnInclude, ->UseQuery - will be deprecated
- Added strict option

0.97 Tue May  8 19:19:00 CEST 2012
- regular release, see changes from dev version below

0.96_005 Sun May  6 16:24:54 CEST 2012
- fix t/25_expr.t number of skip tests

0.96_004 Sat May  5 15:56:27 CEST 2012
- Minor bugfix USE_VARS, module version numbers
- Bugfix caching: under certain circumstances some template includes
  were not added to memory cache and reloaded from file cache very time

0.96_003 Fri May  4 17:09:19 CEST 2012
- Improved Feature: expressions like .path.to.hash{var}

0.96_002 Sun Apr 22 19:23:33 CEST 2012
- Changed/New feature: chomp whitespaces globally or in tags
- New feature: SET_VAR (and USE_VARS)

0.96_001 Sat Apr 21 23:17:59 CEST 2012
- New feature: allow expressions like .path.to.hash{var}
- Bugfix: allow more than two binary operands in H::T::C::Expr

0.96 Mon Nov 21 20:37:22 CET 2011
- fixed 05_filter.t
- fixed t/01_HTML-Template-Compiled.t (problems when test takes longer then
  cache expiration)

0.95_003 Sat Nov 12 15:12:02 CET 2011
- Bugfix: Reload includes of includes (required change of caching mechanism -
  please report any issues if you have problems with caching)
- New Escape: 'IJSON'
- Deprecated the use options 'speed', 'compatible' and 'short'

0.95_002 Mon Oct 31 21:00:51 CET 2011
- HTML::Template Compatibility: lowercase instead of uppercase
  parameters when case_sensitive 0

0.95_001 Sun Aug 28 19:02:00 CEST 2011
- Bugfix: escape=js escapes backslash now (RT 66463)
- Bugfix: H::T::C::Classic ignored case_sensitive (RT 70412)
- Added Template::AutoFilter to bench.pl

0.95 Wed Apr 27 20:48:13 CEST 2011
- Just moving developer release to regular release

0.94_002 Thu Aug 19 23:43:41 CEST 2010
- Fix for 5.12: ignore Storable fatal error abour regexp items.
  thanks to Daniel Tašov for the hint. just a workaround for now

0.94_001 Wed Sep 16 19:41:01 CEST 2009
- Fix memory leak when compiling
  Thanks to Henrik Tougaard

0.94 Thu Sep 10 19:55:04 CEST 2009
- Fix test

0.93_003
- remove HTML::Entities dependency

0.93_002 Fri Aug 21 14:47:09 CEST 2009
- Bugfix: Expressions were only activated when using default
  tagstyle; using tagstyle would overwrite use_expression option
- New feaure: in TMPL_LOOP and TMPL_EACH you can set context="list"
  This is necessary if you call a method that returns a list

examples/bench.pl
#!/usr/bin/perl
# $Id: bench.pl 1126 2011-10-31 19:56:35Z tinita $
use strict;
use warnings;
use Getopt::Long;
use Benchmark qw/ timethese cmpthese /;
use FindBin qw/ $RealBin /;
chdir "$RealBin/..";
#use Devel::Size qw(size total_size);
my $count = 0;
mkdir "cache";
mkdir "cache/htc";
mkdir "cache/htcc";
mkdir "cache/hte";
mkdir "cache/htpl";
mkdir "cache/jit";
my %use = (
	'HTML::Template'           => 0,
    'HTML::Template::Pro'      => 0,
	'HTML::Template::Compiled' => 0,
	'HTML::Template::Pluggable' => 0,
	'HTML::Template::Expr' => 0,
	'HTML::Template::Compiled::Classic' => 0,
#	'HTML::Template::JIT'      => 0,
	'Template'                 => 0,
    'Template::Alloy'                 => 0,
    'Template::HTML'           => 0,
    'Template::AutoFilter'     => 0,
    'Template::Like'           => 0,
	'CGI::Ex::Template'        => 0,
	# not yet
	'Text::ScriptTemplate'     => 0,
    'Text::Xslate'             => 0,
);
for my $key (sort keys %use) {
	eval "require $key";
	$use{$key} = 1 unless $@;
	my $version = $use{$key} ? $key->VERSION : "-";
    printf "using %35s %s\n", $key, $version;
}
HTML::Template::Compiled->clear_filecache("cache/htc");
HTML::Template::Compiled->clear_filecache("cache/htcc");
my $debug = 0;
$ENV{'HTML_TEMPLATE_ROOT'} = "examples";
my $FILE_CACHE     = 0;
my $MEM_CACHE      = 1;
my $LOOP_CONTEXT   = 1;
my $GLOBAL_VARS    = 0;
my $CASE_SENSITIVE = 1;
my $default_escape = 0;
my $template_size  = 1;
my $bench          = 'timethese';
GetOptions(
    "file-cache=i" => \$FILE_CACHE,
    "mem-cache=i" => \$MEM_CACHE,
    "loop-context=i" => \$LOOP_CONTEXT,
    "global-vars=i" => \$GLOBAL_VARS,
    "case-sensitive=i" => \$CASE_SENSITIVE,
    "default-escape=i" => \$default_escape,
    "template-size=i" => \$template_size,
    "bench=s" => \$bench,
);
my $iterations = shift;
my $ht_file = 'test.htc';
my $htcc_file = $ht_file . 'c';
my $tt_file = "test.tt";
my $tst_file = "test.tst";
my $xslfile = "test.xslate";
$template_size =~ tr/0-9//cd;
if ($template_size > 1) {
    for my $file ($ht_file, $htcc_file, $tt_file, $tst_file, $xslfile) {
        open my $fh, "<", "examples/$file" or die "examples/$file: $!";
        my $data = do { local $/; <$fh> };
        my $new_file = "$file.n$template_size";
        open my $out, ">", "examples/$new_file" or die $!;
        print $out $data x $template_size;
        $file = $new_file;
    }
}

print "running with:
--file-cache $FILE_CACHE --mem-cache $MEM_CACHE --loop-context $LOOP_CONTEXT "
. "--global-vars $GLOBAL_VARS --case-sensitive $CASE_SENSITIVE "
. "--default-escape $default_escape --template-size=$template_size
";

my $STDOUT = 0;
sub new_htc {
	my $t1 = HTML::Template::Compiled->new_file( $ht_file,
		#path => 'examples',
		case_sensitive => $CASE_SENSITIVE, # slow down
		loop_context_vars => $LOOP_CONTEXT,
        $default_escape ? (default_escape => 'HTML') : (),
		debug => $debug,
		# note that you have to create the cachedir
		# first, otherwise it will run without cache
#        cache_dir => ($FILE_CACHE ? "cache/htc" : undef),
        file_cache_dir => ($FILE_CACHE ? "cache/htc" : undef),
        file_cache => ($FILE_CACHE ? 1 : undef),

        cache => $MEM_CACHE,
		out_fh => $STDOUT ? 1 : 0,
        global_vars => $GLOBAL_VARS,
        tagstyle => [qw(-asp -comment)],
		expire_time => 1000,
	);
	return $t1;
}

sub new_htcc {
	my $t1 = HTML::Template::Compiled::Classic->new_file( $htcc_file,
		#path => 'examples',
		case_sensitive => $CASE_SENSITIVE, # slow down
		loop_context_vars => $LOOP_CONTEXT,
        $default_escape ? (default_escape => 'HTML') : (),
		debug => $debug,
		# note that you have to create the cachedir
		# first, otherwise it will run without cache
        cache_dir => ($FILE_CACHE ? "cache/htcc" : undef),
        cache => $MEM_CACHE,
		out_fh => $STDOUT ? 1 : 0,
        global_vars => $GLOBAL_VARS,
        debug => 0,
        tagstyle => [qw(-asp -comment)],
		expire_time => 1000,
        #debug => 1,
	);
	return $t1;
}

sub new_tst {
	my $t = Text::ScriptTemplate->new();
	$t->load("examples/$tst_file");
	#my $size = total_size($t1);
	#print "size htc = $size\n";
	return $t;
}

sub new_htp {
	my $t2 = HTML::Template::Pro->new(
		case_sensitive => $CASE_SENSITIVE,
		loop_context_vars => $LOOP_CONTEXT,
        $default_escape ? (default_escape => 'HTML') : (),
		#path => 'examples',
		filename => $ht_file,
#		cache => $MEM_CACHE,
#        $FILE_CACHE ?
#        (file_cache => $FILE_CACHE,
#        file_cache_dir => 'cache/ht') : (),
        global_vars => $GLOBAL_VARS,
#        die_on_bad_params => 0,
	);
	return $t2;
}

sub new_ht {
	my $t2 = HTML::Template->new(
		case_sensitive => $CASE_SENSITIVE,
		loop_context_vars => $LOOP_CONTEXT,
        $default_escape ? (default_escape => 'HTML') : (),
		#path => 'examples',
		filename => $ht_file,
		cache => $MEM_CACHE,
        $FILE_CACHE ?
        (file_cache => $FILE_CACHE,
        file_cache_dir => 'cache/ht') : (),
        global_vars => $GLOBAL_VARS,
        die_on_bad_params => 0,
		blind_cache => 1,
	);
	return $t2;
}
sub new_hte {
	my $t2 = HTML::Template::Expr->new(
		case_sensitive => $CASE_SENSITIVE,
		loop_context_vars => $LOOP_CONTEXT,
        $default_escape ? (default_escape => 'HTML') : (),
		#path => 'examples',
		filename => $ht_file,
		cache => $MEM_CACHE,
        $FILE_CACHE ?
        (file_cache => $FILE_CACHE,
        file_cache_dir => 'cache/hte') : (),
        global_vars => $GLOBAL_VARS,
        die_on_bad_params => 0,
	);
	return $t2;
}
sub new_htpl {
	my $t2 = HTML::Template::Pluggable->new(
		case_sensitive => $CASE_SENSITIVE,
		loop_context_vars => $LOOP_CONTEXT,
        $default_escape ? (default_escape => 'HTML') : (),
		#path => 'examples',
		filename => $ht_file,
		cache => $MEM_CACHE,
        $FILE_CACHE ?
        (file_cache => $FILE_CACHE,
        file_cache_dir => 'cache/htpl') : (),
        global_vars => $GLOBAL_VARS,
        die_on_bad_params => 0,
	);
	return $t2;
}

sub new_htj {
	my $t2 = HTML::Template::JIT->new(
		loop_context_vars => 1,
        $default_escape ? (default_escape => 'HTML') : (),
		#path => 'examples',
		filename => $ht_file,
		cache => 1,
		jit_path => 'cache/jit',
        #global_vars => 1,
	);
	return $t2;
}

sub new_tl {
	my $tt= Template::Like->new(
    );

#        $FILE_CACHE
#            ? (
#                COMPILE_EXT => '.ttc',
#                COMPILE_DIR => 'cache/tt',
#            )
#            : (),
#        $MEM_CACHE
#            ? ()
#            : (CACHE_SIZE => 0),
#		INCLUDE_PATH => 'examples',

	#my $size = total_size($tt);
	#print "size tt  = $size\n";
	return $tt;
}

sub new_tt {
	my $tt= Template->new(
        $FILE_CACHE
            ? (
                COMPILE_EXT => '.ttc',
                COMPILE_DIR => 'cache/tt',
            )
            : (),
        $MEM_CACHE
            ? ()
            : (CACHE_SIZE => 0),
		INCLUDE_PATH => 'examples',
	);
	#my $size = total_size($tt);
	#print "size tt  = $size\n";
	return $tt;
}

sub new_ta {
    my $tt = Template::Alloy->new(
        $FILE_CACHE
            ? (
                COMPILE_EXT => '.ttc',
                COMPILE_DIR => 'cache/tt',
            )
            : (),
        $MEM_CACHE
            ? ()
            : (CACHE_SIZE => 0),
        INCLUDE_PATH => 'examples',
    );
	#my $size = total_size($tt);
	#print "size tt  = $size\n";
    return $tt;
}

sub new_ttaf {
	my $tt= Template::AutoFilter->new(
        $FILE_CACHE
            ? (
                COMPILE_EXT => '.ttc',
                COMPILE_DIR => 'cache/tt',
            )
            : (),
        $MEM_CACHE
            ? ()
            : (CACHE_SIZE => 0),
		INCLUDE_PATH => 'examples',
	);
	#my $size = total_size($tt);
	#print "size tt  = $size\n";
	return $tt;
}

sub new_tth {
	my $tt= Template::HTML->new(
        $FILE_CACHE
            ? (
                COMPILE_EXT => '.ttc',
                COMPILE_DIR => 'cache/tt',
            )
            : (),
        $MEM_CACHE
            ? ()
            : (CACHE_SIZE => 0),
		INCLUDE_PATH => 'examples',
	);
	#my $size = total_size($tt);
	#print "size tt  = $size\n";
	return $tt;
}

sub new_xslate {
    # Text::Xslate has no file caching
    my $t = Text::Xslate->new(
        $MEM_CACHE ? (cache => 2, cache_dir => "cache/xslate") : (),
        path => 'examples',
        syntax => 'TTerse',
    );
    return $t;
}

sub new_cet {
	my $tt= CGI::Ex::Template->new(
        $FILE_CACHE
            ? (
                COMPILE_EXT => '.ttc',
                COMPILE_DIR => 'cache/tt',
            )
            : (),
        $MEM_CACHE
            ? ()
            : (CACHE_SIZE => 0),
		INCLUDE_PATH => 'examples',
	);
	#my $size = total_size($tt);
	#print "size tt  = $size\n";
	return $tt;
}

sub new_st {
	my $st = Text::ScriptTemplate->new;
	$st->load("examples/template.st");
}

my %params = (
	name => '',
	loopa => [{a=>3},{a=>4},{a=>5}],
	#a => [qw(b c d)],
	loopb => [{ inner => 23 }],
	c => [
	 {
		 d=>[({F=>11},{F=>22}, {F=>33})]
	 },
	{
		d=>[({F=>44}, {F=>55}, {F=>66})]
	}
	],
	if2 => 1,
	if3 => 0,
	blubber => "html <test>",
);
open OUT, ">>/dev/null";
#open OUT, ">&STDOUT";

sub output {
	my $t = shift;
	return unless defined $t;
	$t->param(%params);
    if ($STDOUT) {
        my $out;
        if ($t=~m/Compiled/) {
            $out = $t->output(\*OUT);
        }
        else {
            $out = $t->output;
        }
        print OUT $out;
    }
    else {
        my $out = $t->output();
    }
	#print "output():$out\n";
	#my $size = total_size($t);
	#print "size $t = $size\n";
}

sub output_tst {
	my $t = shift;
	return unless defined $t;
	$t->setq(%params,tmpl=>$t);
	my $out = $t->fill;
	#print "output_tst():$out\n";
    if ($STDOUT) {
        print OUT $out;
    }
}
sub output_tl {
	my $t = shift;
	return unless defined $t;
    chdir 'examples';
	my $filett = $tt_file;
    if ($STDOUT) {
        $t->process($filett, \%params, \*OUT) or die $t->error();
    }
    else {
        my $out;
        $t->process($filett, \%params, \$out) or die $t->error();
    }
	#my $size = total_size($t);
	#print "size $t = $size\n";
	#my $out = $t->output;
	#print "\nOUT: $out";
    chdir '..';
}

sub output_tt {
	my $t = shift;
	return unless defined $t;
	my $filett = $tt_file;
	#$t->process($filett, \%params, \*OUT);
    if ($STDOUT) {
        $t->process($filett, \%params, \*OUT) or die $t->error();
    }
    else {
        my $out;
        $t->process($filett, \%params, \$out) or die $t->error();
    }
	#my $size = total_size($t);
	#print "size $t = $size\n";
	#print $t->{code} if exists $t->{code};
    #my $out = $t->output;
    #print "\nOUT: $out";
}

sub output_ttaf {
	my $t = shift;
	return unless defined $t;
	my $filett = $tt_file;
	#$t->process($filett, \%params, \*OUT);
    if ($STDOUT) {
        $t->process($filett, \%params, \*OUT) or die $t->error();
    }
    else {
        my $out;
        $t->process($filett, \%params, \$out) or die $t->error();
    }
	#my $size = total_size($t);
	#print "size $t = $size\n";
	#print $t->{code} if exists $t->{code};
    #my $out = $t->output;
    #print "\nOUT: $out";
}

sub output_xslate {
	my $t = shift;
	return unless defined $t;
	my $filett = $tt_file;
	#$t->process($filett, \%params, \*OUT);
#	$t->process($filett, \%params, \*OUT) or die $t->error();
	#my $size = total_size($t);
	#print "size $t = $size\n";
	#print $t->{code} if exists $t->{code};
    my $out = $t->render($xslfile, \%params);
    #my $out = $t->output;
    if ($STDOUT) {
        print OUT $out;
    }
}

my $global_htc = $use{'HTML::Template::Compiled'} ? new_htc : undef;
my $global_htcc = $use{'HTML::Template::Compiled::Classic'} ? new_htcc : undef;
my $global_ht = $use{'HTML::Template'} ? new_ht : undef;
my $global_htp = $use{'HTML::Template::Pro'} ? new_htp : undef;
my $global_htpl = $use{'HTML::Template::Pluggable'} ? new_htpl : undef;
my $global_htj = $use{'HTML::Template::JIT'} ? new_htj : undef;
my $global_tt = $use{'Template'} ? new_tt : undef;
my $global_ta = $use{'Template::Alloy'} ? new_ta : undef;
my $global_ttaf = $use{'Template::AutoFilter'} ? new_ttaf : undef;
my $global_tth = $use{'Template::HTML'} ? new_tth : undef;
my $global_xslate = $use{'Text::Xslate'} ? new_xslate : undef;
my $global_tl = $use{'Template::Like'} ? new_tl : undef;
my $global_cet = $use{'CGI::Ex::Template'} ? new_cet : undef;
my $global_tst = $use{'Text::ScriptTemplate'} ? new_tst : undef;
if(1) {
my %args = (
       $use{'HTML::Template::Compiled'} ? (
            # deactivate memory cache
            #new_htc_w_clear_cache => sub {my $t = new_htc();$t->clear_cache},
            # normal, with memory cache
            # new_htc => sub {my $t = new_htc()},
            #output_htc => sub {output($global_htc)},
            all_htc => sub {my $t = new_htc();output($t)},
        ) : (),
        $use{'HTML::Template::Compiled::Classic'} ? (
            # deactivate memory cache
            #new_htc_w_clear_cache => sub {my $t = new_htc();$t->clear_cache},
            # normal, with memory cache
            # new_htcc => sub {my $t = new_htcc()},
            #output_htc => sub {output($global_htc)},
            all_htcc => sub {my $t = new_htcc();output($t)},
        ) : (),
		$use{'HTML::Template'} ? (
            # new_ht => sub {my $t = new_ht()},
			#output_ht => sub {output($global_ht)},
						all_ht => sub {my $t = new_ht();output($t)},
        ) : (),
		$use{'HTML::Template::Pro'} ? (
            # new_htp => sub {my $t = new_htpl()},
			#output_htp => sub {output($global_htp)},
						all_htp => sub {my $t = new_htp();output($t)},
        ) : (),
		$use{'HTML::Template::Pluggable'} ? (
            # new_htpl => sub {my $t = new_htpl()},
			#output_htpl => sub {output($global_htpl)},
						all_htpl => sub {my $t = new_htpl();output($t)},
        ) : (),
        $use{'HTML::Template::Expr'} && !$FILE_CACHE ? (
            # new_hte => sub {my $t = new_hte()},
            #output_hte => sub {output($global_hte)},
            all_hte => sub {my $t = new_hte();output($t)},
        ) : (),
        $use{'HTML::Template::JIT'} ? (
					#new_htj => sub {my $t = new_htj();},
					#output_htj => sub {output($global_htj)},
						all_htj => sub {my $t = new_htj();output($t)},
        ) : (),
        $use{'Template'} ? (
            #new_tt => sub {my $t = new_tt();},
            #output_tt => sub {output_tt($global_tt)},
            process_tt => sub {output_tt($global_tt)},
            $MEM_CACHE
                ? ()
                : (all_tt_new_object => sub {my $t = new_tt();output_tt($t)}),
        ): (),
        $use{'Template::Alloy'} ? (
            #new_tt => sub {my $t = new_tt();},
            #output_tt => sub {output_tt($global_tt)},
            process_ta => sub {output_tt($global_ta)},
            $MEM_CACHE
                ? ()
                : (all_ta_new_object => sub {my $t = new_ta();output_tt($t)}),
        ): (),
        $use{'Template::AutoFilter'} ? (
            #new_ttaf => sub {my $t = new_ttaf();},
            #output_ttaf => sub {output_ttaf($global_ttaf)},
            process_ttaf => sub {output_ttaf($global_ttaf)},
            $MEM_CACHE
                ? ()
                : (all_ttaf_new_object => sub {my $t = new_ttaf();output_ttaf($t)}),
        ): (),
        $use{'Template::HTML'} ? (
            #new_tt => sub {my $t = new_tt();},
            #output_tt => sub {output_tt($global_tt)},
            process_tth => sub {output_tt($global_tth)},
            $MEM_CACHE
                ? ()
                : (all_tth_new_object => sub {my $t = new_tth();output_tt($t)}),
        ): (),
        $use{'Text::Xslate'} ? (
            #new_tt => sub {my $t = new_tt();},
            #output_tt => sub {output_tt($global_tt)},
            process_xslate => sub {output_xslate($global_xslate)},
            $MEM_CACHE
                ? ()
                : (all_xslate_new_object => sub {my $t = new_xslate();output_xslate($t)}),
        ): (),
        $use{'Template::Like'} ? (
            process_tl => sub {output_tl($global_tl)},
        ): (),
        $use{'CGI::Ex::Template'} ? (
            #new_tt => sub {my $t = new_tt();},
            #output_tt => sub {output_tt($global_tt)},
            process_cet => sub {output_tt($global_cet)},
            $MEM_CACHE
                ? ()
                : (all_cet_new_object => sub {my $t = new_cet();output_tt($t)}),
        ): (),
        $use{'Text::ScriptTemplate'} ? (
					#new_tst => sub {my $t = new_tst();},
                    #output_tst => sub {output_tst($global_tst)},
						all_tst => sub {my $t = new_tst();output_tst($t)},
        ): (),
);
# try to align table correctly also for longer strings
my %args_new;
for my $key (keys %args) {
    my $new_key = sprintf "%21s", $key;
    $args_new{ $new_key } = $args{ $key };
}
if ($bench eq 'timethese') {
    timethese ($iterations||-1, {
        %args_new
 	});
}
elsif ($bench eq 'cmpthese') {
    cmpthese ($iterations||-1, {
        %args_new
 	});
}
}
__END__
